File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
# 12865 평범한 배낭
import sys

N, K = map(int, sys.stdin.readline().split())

dp = [[None] * (N + 1) for _ in range(K + 1)]
for items in range(0, N + 1):
    dp[0][items] = 0

for back in range(0, K + 1):
    dp[back][0] = 0

item = [[] for _ in range(N + 1)]
for it in range(1, N + 1):
    weight, value = map(int, sys.stdin.readline().split())
    item[it] = [weight, value]

WEIGHT = 0
VALUE = 1


def dynamic(in_item, back_weight):
    if dp[back_weight][in_item] is not None:
        return dp[back_weight][in_item]

    # if in_item == 0 or back_weight == 0:
    #     return 0
    if back_weight < item[in_item][0]:
        dp[back_weight][in_item] = dynamic(in_item - 1, back_weight)
    else:
        dp[back_weight][in_item] = (
            max(dynamic(in_item - 1, back_weight),
                dynamic(in_item - 1, back_weight - item[in_item][WEIGHT]) + item[in_item][VALUE]))

    return dp[back_weight][in_item]


print(dynamic(N, K))

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
//  평범한 배낭

import java.util.*;
import java.lang.*;
import java.io.*;

class Main {

    private static int[][] dp;
    private static myObject[] prob;
    private static int K;
    private static int N;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] line = br.readLine().split(" ");
        N = Integer.parseInt(line[0]);
        K = Integer.parseInt(line[1]);

        dp = new int[N + 1][K + 1];

        prob = new myObject[N + 1];
        for (int i = 1; i < N + 1; i++) {
            line = br.readLine().split(" ");
            prob[i] = new myObject(Integer.parseInt(line[0]), Integer.parseInt(line[1]));
        }

        System.out.println(dynamic(N,K));
    }

    private static int dynamic(int count, int weight) {
        if (dp[count][weight] != 0) {
            return dp[count][weight];
        }

        if (count == 0) return 0;

        if (prob[count].weight <= weight) {
            dp[count][weight] = Math.max(
                    dynamic(count - 1, weight), dynamic(count - 1, weight - prob[count].weight) + prob[count].value
            );
        } else {
            dp[count][weight] = dynamic(count - 1, weight);
        }

        return dp[count][weight];
    }

    private static class myObject{
        int weight;
        int value;

        public myObject(int weight, int value) {
            this.weight = weight;
            this.value = value;
        }
    }
}
